File: csharp/sdk/src/ModelContextProtocol.Interceptors/Gateway/InterceptorChainRunner.cs
using System.Text.Json.Nodes;
using ModelContextProtocol.Client;
using ModelContextProtocol.Interceptors.Client;
using ModelContextProtocol.Interceptors.Protocol;

namespace ModelContextProtocol.Interceptors.Gateway;

/// <summary>
/// Encapsulates the logic for running interceptor chain phases (request/response)
/// against one or more interceptor server clients. Used by both <see cref="Client.InterceptingMcpClient"/>
/// and <see cref="McpInterceptorGateway"/>.
/// </summary>
internal sealed class InterceptorChainRunner
{
    private readonly IReadOnlyList<McpClient> _interceptorClients;
    private readonly IList<string>? _events;
    private readonly int? _timeoutMs;
    private readonly InvokeInterceptorContext? _defaultContext;

    internal InterceptorChainRunner(
        IReadOnlyList<McpClient> interceptorClients,
        IList<string>? events = null,
        int? timeoutMs = null,
        InvokeInterceptorContext? defaultContext = null)
    {
        _interceptorClients = interceptorClients;
        _events = events;
        _timeoutMs = timeoutMs;
        _defaultContext = defaultContext;
    }

    /// <summary>
    /// Returns whether the given event should be intercepted based on the configured event filter.
    /// </summary>
    internal bool ShouldIntercept(string eventName)
    {
        if (_events is not { Count: > 0 } events)
        {
            return true;
        }

        return events.Contains(eventName);
    }

    /// <summary>
    /// Runs the interceptor chain phase across all configured interceptor clients as a single
    /// merged chain per the SEP: interceptors discovered from every client are combined, sorted
    /// globally by priority hint (alphabetical tie-break), and each is invoked on the client that
    /// hosts it via <see cref="InterceptorChain.ExecuteAsync(IEnumerable{McpClient}, ExecuteChainRequestParams, CancellationToken)"/>.
    /// </summary>
    /// <returns>The final payload (the original on failure) and the chain status.</returns>
    internal async ValueTask<(JsonNode payload, InterceptorChainStatus status)> RunChainPhaseAsync(
        string eventName, InterceptorPhase phase, JsonNode payload, CancellationToken ct)
    {
        var chainResult = await InterceptorChain.ExecuteAsync(
            _interceptorClients,
            new ExecuteChainRequestParams
            {
                Event = eventName,
                Phase = phase,
                Payload = payload,
                TimeoutMs = _timeoutMs,
                Context = _defaultContext,
            },
            ct);

        return chainResult.Status == InterceptorChainStatus.Success
            ? (chainResult.FinalPayload ?? payload, InterceptorChainStatus.Success)
            : (payload, chainResult.Status);
    }

    /// <summary>
    /// Throws an appropriate exception for a non-success chain status.
    /// </summary>
    internal static void ThrowChainFailure(string operation, InterceptorPhase phase, InterceptorChainStatus status)
    {
        var phaseText = phase == InterceptorPhase.Request ? "Request" : "Response";
        if (status == InterceptorChainStatus.ValidationFailed)
        {
            throw new McpInterceptorValidationException($"{phaseText}-phase interceptor validation failed for {operation}.");
        }

        throw new InvalidOperationException($"{phaseText}-phase interceptor chain failed for {operation} with status '{status}'.");
    }
}

File: typescript/sdk/src/client/interceptor-chain-runner.ts
// Copyright 2025 The MCP Interceptors Authors. All rights reserved.
import type { Client } from "@modelcontextprotocol/client";

import { throwChainFailure } from '../protocol/errors.js';
import type {
  ExecuteChainRequestParams,
  InterceptorChainResult,
  InterceptorOverrides,
  InterceptorPhase,
  InvokeInterceptorContext,
} from '../protocol/types.js';
import { matchesEvent } from './chain-orchestrator.js';
import { executeInterceptorChainOnClients } from './execute-interceptor-chain-on-clients.js';
import type { DuplicateInterceptorNamePolicy } from './interceptor-chain-entry.js';

export interface InterceptorChainRunnerOptions {
  interceptorClients: Client[];
  /** When set, only these lifecycle events are intercepted. `'*'` intercepts every event. */
  events?: string[];
  timeoutMs?: number;
  defaultContext?: InvokeInterceptorContext;
  /** Passed to multi-host chain merge when more than one interceptor client is configured. */
  duplicateNamePolicy?: DuplicateInterceptorNamePolicy;
  /** Invoker-side execution policy per interceptor name (SEP chain-entry overrides). */
  overrides?: Record<string, InterceptorOverrides>;
}

export class InterceptorChainRunner {
  private readonly clients: Client[];
  private readonly events: string[] | undefined;
  private readonly timeoutMs: number | undefined;
  private readonly defaultContext: InvokeInterceptorContext | undefined;
  private readonly duplicateNamePolicy: DuplicateInterceptorNamePolicy | undefined;
  private readonly overrides: Record<string, InterceptorOverrides> | undefined;

  constructor(options: InterceptorChainRunnerOptions) {
    this.clients = options.interceptorClients;
    this.events = options.events;
    this.timeoutMs = options.timeoutMs;
    this.defaultContext = options.defaultContext;
    this.duplicateNamePolicy = options.duplicateNamePolicy;
    this.overrides = options.overrides;
  }

  shouldIntercept(eventName: string): boolean {
    if (!this.events || this.events.length === 0) {
      return true;
    }
    return matchesEvent(this.events, eventName);
  }

  async runChainPhase(
    eventName: string,
    phase: InterceptorPhase,
    payload: unknown,
    signal?: AbortSignal,
  ): Promise<{ payload: unknown; chainResult: InterceptorChainResult }> {
    const currentPayload = payload;

    const chainParams: ExecuteChainRequestParams = {
      event: eventName,
      phase,
      payload: currentPayload,
      timeoutMs: this.timeoutMs,
      context: this.defaultContext,
    };

    const chainResult = await executeInterceptorChainOnClients(
      this.clients.map((client, index) => ({ client, label: `host-${index}` })),
      chainParams,
      { signal, duplicateNamePolicy: this.duplicateNamePolicy, overrides: this.overrides },
    );

    if (chainResult.status !== 'success') {
      return { payload: currentPayload, chainResult };
    }

    return {
      payload: chainResult.finalPayload ?? currentPayload,
      chainResult,
    };
  }

  async runChainPhaseOrThrow(
    operation: string,
    eventName: string,
    phase: InterceptorPhase,
    payload: unknown,
    signal?: AbortSignal,
  ): Promise<unknown> {
    const { payload: processed, chainResult } = await this.runChainPhase(
      eventName,
      phase,
      payload,
      signal,
    );
    if (chainResult.status !== 'success') {
      throwChainFailure(operation, phase, chainResult.status, chainResult);
    }
    return processed;
  }
}
